AppleScript support

Wevah opened the pull request:
WIP for issue #316.

I tried to make as few changes to the pre-existing files as possible; most stuff is in extensions in separate files. The only existing API changes were adding parent references to tracks and playlist items. Scripting needs those for returning object specifiers, and they're also used to make sure that actions like trying to set a current track to a track belonging to another player is a no-op or throws an error. (An alternative would be to iterate every player/track or player/playlist item combo, but that felt inefficient.)

`unowned` seemed appropriate since presumably if the player goes away, so do all the tracks/playlist items. Can be changed to `weak` if necessary.

Wevah committed:
Preliminary scripting support

Wevah committed:
Make FourCharCode conform to ExpressibleByStringLiteral

Wevah committed:
Use a private extension on FourCharCode for constants

Wevah committed:
Do the bit math ourselves instead of using NSHFSTypeCodeFromFileType()

Wevah committed:
Add 10.12+ message to pip property; update seeking description

Wevah committed:
Make sure tracks belong to the player

Wevah committed:
Add error messages for setting the current track

Wevah committed:
Add support for next and previous

Wevah commented:
One other thing I wanted to add: I tried to mimic the Music/iTunes sdef reasonably closely, so I used the same `FourCharCode`s for equivalent commands/properties, and followed the FCC naming prefix convention (`e` for enum, `p` for property, etc.).

Wevah commented:
Hit close by accident.

saagarjha commented on iina/AppDelegate+Scripting.swift:
```diff
@@ -0,0 +1,34 @@
+extension AppDelegate {
+
+  func application(_ sender: NSApplication, delegateHandlesKey key: String) -> Bool {
+    return Set(["orderedPlayerWindows", "orderedPlayers"]).contains(key)
```
For this few items, this could probably just be an explicit check–i.e. `key == "orderedPlayerWindows" || key == "orderedPlayerWindows"`. Unless you think there's a good reason to not do it this way?

Wevah replied:
Probably just a holdover from code I had in another app with a *lot* more matches (to reduce noise).

Wevah replied:
Actually, `orderedPlayerWindows` isn't even a thing currently, so I'll remove it.

saagarjha commented on iina/AppDelegate+Scripting.swift:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  @objc var orderedPlayers: [PlayerCore] {
+    let windows = NSApp.orderedWindows
+    var players = [PlayerCore]()
+
+    for window in windows {
+      if window.isVisible, let controller = window.delegate as? PlayerWindowController {
+        players.append(controller.player)
+      }
+    }
+
+    return players
```
This could all probably be done functionally, e.g.
```
Array(NSApp.orderedWindows.lazy.filter(\.isVisible).flatMap { $0.delegate as? PlayerWindowController}.map(\.player))
```

saagarjha commented on iina/FourCharCode+Extensions.swift:
```diff
@@ -0,0 +1,25 @@
+
+  public init(stringLiteral: String) {
+    // Match NSHFSTypeCodeFromFileType() behavior: Return 0 for invalid codes.
+    guard stringLiteral.count == 4, let data = stringLiteral.data(using: .macOSRoman) else {
```
I think it might be a good idea to have initialization fail if these requirements aren't met. What do you think?

Wevah replied:
I’m fine doing it that way; that was my initial thought, but then I decided to match the behavior of the notes function.

Wevah replied:
I'd have to make it throw, since a failable initializer can't satisfy the protocol.

saagarjha commented on iina/MPVPlaylistItem+Scripting.swift:
```diff
@@ -0,0 +1,48 @@
+
+extension MPVPlaylistItem {
+
+  @objc var scriptingURL: String { isNetworkResource ? filename : URL(fileURLWithPath: filename).absoluteString }
```
I think naming this "URL" might a bit confusing if it's not actually a URL.

Wevah replied:
Good call; I’ll change it to `scriptingURLString` or something.

saagarjha reviewed (commented):
This is great; I think this can really help us towards our goal of making IINA extensible. I'll need to do a bit of testing with the scripting stuff to see how this works, but for now here's a couple of general comments that I had while reading the code.

Wevah committed:
sciptingURL -> scriptingURLString

Wevah committed:
"orderedPlayerWindows" isn't implemented currently

Wevah committed:
Be functional in orderedPlayers

Wevah commented:
More notes as I remember them:

- `AppDelegate` has a stub for `handlePlayCommand:`; it'll probably just forward to the frontmost player.
- `play` could take a file or URL, but using `open` might be better? Cocoa scripting conflates the direct parameter and the command handler without some AppleEvents property finagling.

Wevah committed:
Remove unneeded @objc parameter

Wevah commented:
Looks like the build failed due to using key paths as functions and Travis not having an Xcode 11.4 image yet.

Wevah committed:
Fix an incorrect FourCharCode

Wevah committed:
Set @objc on extensions instead of each individual property/method

Wevah commented:
Is there anything that looks missing from here that should be in for a first go?

hfpf commented:
Why leave the WIP label here?

sclsj commented:
@Wevah I forked your fork and merged the applescript branch. I can now open the sdef in `Script Editor`, but doing anything results in no actions (e.g. playpause, play, and pause don't work). Is this normal? My understanding is that your code works? Can you take a look if I'm doing anything wrong? https://github.com/sclsj/iina.

Wevah commented:
@sclsj Hm, the only things you should need are the .sdef, the files with the extensions on various classes and the minor tweaks to those classes (I think I added a referential property to at least one of them).

Wevah commented:
> Why leave the WIP label here?

Good call; removed.

Wevah commented:
Well this is annoying, the xcworkspace file opens empty in the latest Xcode I can run on this old machin.

Wevah commented:
Looks like I accidentally wipted out the `NSAppleScriptEnabled` Info.plist key when I rebased.

Wevah commented:
> @Wevah I forked your fork and merged the applescript branch. I can now open the sdef in `Script Editor`, but doing anything results in no actions (e.g. playpause, play, and pause don't work). Is this normal? My understanding is that your code works? Can you take a look if I'm doing anything wrong? https://github.com/sclsj/iina.

I think I see the issue; you need to tell an individual player to playpause, etc., since conceivably someone could have multiple players open:

```applescript
tell application "IINA"
	tell front player to playpause
end tell
```

Maybe I should add an app-wide version that automatically forwards to the first player?

Wevah committed:
Fix misnaming of next/previous item

Wevah commented:
Now I remember why it was marked as WIP still; I missed a few things.

Wevah committed:
Add default implementations for playback commands

…and move scripting stuff to a subfolder.

Wevah committed:
Fix direct parameter description

sclsj commented:
@Wevah How would I pause the currently playing player and start the last playing player?

Wevah committed:
Rename playing property on playlist item to avoid conflict with the state constant

Wevah commented:
@sclsj You'd probably have to store a reference to the active player, or the player's id, somewhere if you've got multiple players open.

low-batt commented:
Some feedback from another IINA user...
Nice work!!

I successfully built the PR natively on an Apple Silicon Mac running Monterey 12.0.1 with mpv 0.34.0 and FFmpeg 4.4.1. I ran several AppleScript tests and they passed. Yeah!

I am not an AppleScript expert, so I could be wrong about some of the issues I'm about to bring up. See what others think.

This new warning is being emitted. Whatever is causing it needs to be fixed:
```text
2021-11-21 17:19:18.691940-0500 IINA[44532:1033619] .sdef warning for argument 'FileType' of command 'save' in suite 'Standard Suite': 'saveable file format' is not a valid type name.
```
IINA's support for AppleScript should follow Apple standards as much as possible. In particular we want to insure the AppleScript capabilities provided by the Cocoa Framework work as expected. This shows a problem:
```text
low-batt@gag ~$ osascript -e 'tell application "IINA"' -e 'activate' -e 'open "/Users/low-batt/Movies/issue-316.mp4"' -e 'end tell'
missing value
```
Running the same script against QuickTime shows what is wrong:
```text
low-batt@gag ~$ osascript -e 'tell application "QuickTime Player"' -e 'activate' -e 'open "/Users/low-batt/Movies/issue-316.mp4"' -e 'end tell'
document issue-316.mp4
```
The IINA.sdef contains:
```xml
<xi:include href="file:///System/Library/ScriptingDefinitions/CocoaStandard.sdef" xpointer="xpointer(/dictionary/suite)"/>
```
Which is fine if you fully follow the Cocoa conventions. These standard definitions expect that `NSDocument` is being used. QuickTime is using documents. It makes sense for QuickTime because it is not just a player. It has operations like `trim` that make modifications that require saving, whereas IINA is strictly a player. Need to consider now if IINA will in the future add features that involve modifying and saving files. If so, then some thought should be given as to whether IINA should be using `NSDocument`. It is not hard to add that support.

The Cocoa Standard definition in question is:
```xml
<command name="open" code="aevtodoc" description="Open a document.">
  <direct-parameter description="The file(s) to be opened.">
    <type type="file"/>
    <type type="file" list="yes"/>
  </direct-parameter>
  <result description="The opened document(s).">
    <type type="document"/>
    <type type="document" list="yes"/>
  </result>
</command>
```
The open command **must** return a document if that definition is used. IINA must either adopt the Cocoa QuickTime "document" pattern or address this issue in another manner.

Looking at the definitions using the Script Editor:

![issue-316-quicktime](https://user-images.githubusercontent.com/86170219/142798188-4846d722-996e-4f12-8cfa-8b2a56691899.jpg)

![issue-316-iina](https://user-images.githubusercontent.com/86170219/142798959-dc6c73f1-6bc0-49cd-845f-f93766b31a9d.jpg)

The QuickTime definitions contains an "Internet Suite" with an "Open URL" command., would be nice if IINA had this.

Would be nice for IINA to have commands for "Next Frame" and "Previous Frame".

QuickTime has "audio volume" whereas IINA has "sound volume", the term used by the Music app . Since IINA refers to this as "audio / volume" in the menus the AppleScript definitions should match up with the naming used in the menus. Would also be good to provide some sort of indication of what values are expected like the QuickTime definitions do.

Apple's AppleScript naming guidelines stress the importance of not picking a name for a property that makes it seem like a command. I expect that is why the QuickTime property is named "muted". The IINA property is named "mute", likely coming from the Music app which failed to follow the guidelines and has a property named "mute".

IINA has a "pause" command, would be nice to have a "resume" command like QuickTime.

Wevah committed:
Set Safari extension target OS to 10.12 to defeat errors

Wevah committed:
Add rotation scripting property

Also use contains instead of firstIndex(of:) when testing for a valid rotation.

Wevah commented:
> The IINA.sdef contains:
> 
> ```
> <xi:include href="file:///System/Library/ScriptingDefinitions/CocoaStandard.sdef" xpointer="xpointer(/dictionary/suite)"/>
> ```

Good call; I should probably get rid of that.

> The QuickTime definitions contains an "Internet Suite" with an "Open URL" command., would be nice if IINA had this.
> 
> Would be nice for IINA to have commands for "Next Frame" and "Previous Frame".

Reasonable!

> QuickTime has "audio volume" whereas IINA has "sound volume", the term used by the Music app . Since IINA refers to this as "audio / volume" in the menus the AppleScript definitions should match up with the naming used in the menus. Would also be good to provide some sort of indication of what values are expected like the QuickTime definitions do.
> 
> Apple's AppleScript naming guidelines stress the importance of not picking a name for a property that makes it seem like a command. I expect that is why the QuickTime property is named "muted". The IINA property is named "mute", likely coming from the Music app which failed to follow the guidelines and has a property named "mute".

Yes, I've been going off of the Music app's names and type codes. I suppose mimicking QuickTime would make more sense.

> IINA has a "pause" command, would be nice to have a "resume" command like QuickTime.

Does "resume" do something different in QuickTime than calling "play" on a paused player in IINA? I guess I should check.

Wevah commented:
In retrospect, I maybe shouldn't have rebased. Oh well.

Wevah committed:
Add mirror/flip properties and next/previous frame commands

Wevah commented:
If I'm reading QuickTime's dictionary right, `resume` is for resuming recordings, not playback.

Wevah committed:
Fix typo

Wevah commented:
I'm going to add a custom pruned copy of the core suite instead of importing it (like Xcode does).

Wevah committed:
Fix exception when built using Xcode 13

Wevah committed:
Use our own copy of the standard suite

Wevah committed:
Make players respond to the close command

Wevah committed:
Add a range description to the audio volume property

Wevah committed:
Remove unnecessary constraint

Fixes a bunch of layout warnings.

CarterLi commented:
Hello! I'm interested in merging this PR into my [iina-plus](https://github.com/iina-plus/iina) repo, which tries to collect various awesome features into one build.

However I got some conflict errors when merging

```
$ git merge Wevah/applescript
Auto-merging iina.xcodeproj/project.pbxproj
CONFLICT (content): Merge conflict in iina.xcodeproj/project.pbxproj
Auto-merging iina/Base.lproj/PrefGeneralViewController.xib
CONFLICT (content): Merge conflict in iina/Base.lproj/PrefGeneralViewController.xib
Auto-merging iina/Info.plist
CONFLICT (content): Merge conflict in iina/Info.plist
Auto-merging iina/PlayerCore.swift
Auto-merging iina/PrefGeneralViewController.swift
Automatic merge failed; fix conflicts and then commit the result.
```

As it doesn't seems to be merged into the official repo soon, would you like to open a PR to my iina-plus repo?

ddddxxx commented:
Hi @Wevah, I want to communicate with IINA via AppleScript in my app. This pr seems very promising and I’d like to work on it right now. I’m wondering if there will be major changes in `sdef` file. Small changes like renaming a property is acceptable. Or I need to wait a bit longer for the script interface to stabilize?

Wevah commented:
@ddddxxx It looks like this got merged into `develop`, so there probably won't be many more changes to the functionality as it exists now unless something breaks,

low-batt commented:
Busy today, having to write this in a rush...

@ddddxxx Here is my own opinion on your question:

All bets are off until this feature is in an official released version of IINA. Program against the current proposed interface at your own risk.

The reason I say that is because of the issue you are raising. Unlike other IINA features where the user interface can be "tweaked" in future releases and humans can adjust to it, this feature requires that backward compatibility with programs and scripts be taken into account when making any future changes. For that reason this feature requires an intense detailed picky peer review as well as feedback from Beta test users. Since that has not been completed there is no guarantee that the interface will not change.

@Wevah has done some great work here. It is on my list to take a close look at this but I've been distracted, trying to fix other issues.

**NOTE** that this has **NOT** been merged into IINA yet. That is why this PR is still open. The merge notice seen above is because CarterLi merged it into his fork, "IINA-Plus".

Wevah commented:
Ahh, whoops! I must be blind.

CarterLi commented:
> ... this feature requires that backward compatibility with programs and scripts be taken into account when making any future changes. For that reason this feature requires an intense detailed picky peer review as well as feedback from Beta test users.

You are right. However I don't think there are many *Beta test users*. No one can tell if an API is stable without actually using it ( implementing something based on it ). It can be a good time to prove this API is stable enough to use or not.
